kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+        let (process_option, unused_memory) = unsafe {
+            ProcessStandard::create(
+                self.kernel,
+                self.chip,
+                process_binary,
+                app_memory,
+                self.fault_policy,
+                app_id,
+                index,
+            )
+            .map_err(|(e, memory)| (memory, e))?
+        };
+
+        process_option.map(|process| {
+            if config::CONFIG.debug_load_processes {
+                debug!(
+                    "Loading: {} [{}] flash={:#010X}-{:#010X} ram={:#010X}-{:#010X}",
+                    process.get_process_name(),
+                    index,
+                    process.get_addresses().flash_start,
+                    process.get_addresses().flash_end,
+                    process.get_addresses().sram_start,
+                    process.get_addresses().sram_end - 1,
+                );
+            }
+        });
+        Ok((unused_memory, process_option))
+    }
+}
+
+/// This is the callback client for the underlying physical storage driver.
+impl<'a, C: 'static + Chip> NonvolatileStorageClient for DynamicProcessLoader<'a, C> {
+    fn read_done(&self, _buffer: &'static mut [u8], _length: usize) {
+        //we will never use this, but we need to implement this anyway
+        unimplemented!();
+    }
+
+    fn write_done(&self, buffer: &'static mut [u8], length: usize) {
+        // Switch on which user generated this callback.
+        self.current_user.take().map(|user| {
+            match user {
+                NonvolatileUser::Client => {
+                    // trigger the client callback
+                    self.client.map(|client| {
+                        client.app_data_write_done(buffer, length);
+                    });
+                }
+                NonvolatileUser::Kernel => {
+                    // replace the buffer after the padding is written
+                    self.buffer.replace(buffer);
+                }
+            }
+        });
+    }
+}
+
+// Interface exposed to the app_loader capsule
+impl<'a, C: 'static + Chip> DynamicProcessLoading for DynamicProcessLoader<'a, C> {
+    fn set_client(&self, client: &'static dyn DynamicProcessLoadingClient) {
+        self.client.set(client);
+    }
+
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode> {
+        //TODO(?): Check if it is a newer version of an existing app.
+        // We can potentially flash the new app and load it before erasing the old one.
+        // What happens to the process though? Need to delete it from the process array and load it back in?
+
+        let mut new_app_data = NewApp {
+            // struct to hold some information about the new app
+            size: 0,
+            start_addr: 0,
+        };
+
+        let flash_start = self.flash.get().as_ptr() as usize; //start of the flash region
+
+        new_app_data.start_addr = self.flash.get().as_ptr() as usize;
+        new_app_data.size = app_length;
+
```

Comment:
I think this should check that an app isn't already in the process of being loaded.

Reply:
The state tracking implementation ensures this does not happen. I've tested it on my current build that I just pushed.